test/CodeGen/AVR/umul.with.overflow.i16-bug.ll

```diff
@@ -0,0 +1,52 @@
+; RUN: llc -O1 < %s -march=avr | FileCheck %s
+
+target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-apple-macosx10.9"
+
+%Vs6UInt16 = type <{ i16 }>
+%Sb = type <{ i1 }>
+
+define hidden void @_TF4main13setServoAngleFT5angleVs6UInt16_T_(i16) #0 {
+  ; CHECK-LABEL: entry
+entry:
+  %adjustedAngle = alloca %Vs6UInt16, align 2
+  %1 = bitcast %Vs6UInt16* %adjustedAngle to i8*
+  %adjustedAngle._value = getelementptr inbounds %Vs6UInt16, %Vs6UInt16* %adjustedAngle, i32 0, i32 0
+  store i16 %0, i16* %adjustedAngle._value, align 2
+
+;print(unsignedInt: adjustedAngle &* UInt16(11))
+; breaks here
+  %adjustedAngle._value2 = getelementptr inbounds %Vs6UInt16, %Vs6UInt16* %adjustedAngle, i32 0, i32 0
+  %2 = load i16, i16* %adjustedAngle._value2, align 2
+
+; CHECK: mov r22, r24
```

Comment:
I recommend using regexes for the temporary registers so that the test is resilient against regalloc changes

This can be done like 

```asm
; CHECK: movw [[REG1:(r[0-9]+|[XYZ])]], r24
; CHECK: movw [[REG2:(r[0-9]+|[XYZ])]], r22
; CHECK: movw r22, r18
; CHECK: movw r24, r20
; CHECK: movw r18, [[REG2]]
; CHECK: movw r20, [[REG1]]
```

Reply:
Fixed registers like the function arguments in `r24:25`  are fine to leave hardcoded though.

Reply:
I'm actually thinking, looking at the above code, that I need to see what happens when the patches for avr-rust#130 and avr-rust#129 are combined together, because they may interact and change the code produced. After that, a patch to the above test might be necessary.

Reply:
OK, I just re-tested it with both patches and all unit tests pass. Plus I went back and re-read the assembly language carefully and all registers have to be exactly as specified.  When r24/r25 are moved to r22/r23, that's part of the setup for the call, r22/r23 aren't temp registers, they're part of the ABI of __mulsi3.  They're the lower two bytes of a 32 bit parameter. The upper two bytes (r24/r25) of the parameter are copied from r20, r21 but could just as easily have been set to 0 directly.  The overall effect is the function takes an int16, zero extends it to 32 bits then multiplies by 11 in a 32-bit parameter, and prints the bottom 16 bits of the result. I think the existing unit test is exactly right. :)

Reply:
Such a well written explanation that I automatically started skim reading with the presumption of truth until realizing that I should actually absorb it.

I agree on all points, sorry for the lame delay!